index.py

```diff
@@ -27,24 +27,44 @@ def restore_index(self):
             logging.info('No previous index was found')
 
     def search(self, query: str, number_of_results: int):
-        if len(query.split(' ')) > 1:
-            raise AttributeError('Only single word query\'s are supported for now')
         processed_query = self.get_tokens(query)
-        results = list()
+
+        result_stack = []
+        pending_operation = ''
         for token in processed_query:
-            pointers = self.index.get(token)
-            if not pointers:
-                continue
-            i = 0
-            for pointer in pointers:
-
-                comment = CsvHelper.read_comment(pointer)
-                results.append(comment)
-                if i > number_of_results:
-                    break
-                i += 1
-
-        return results
+            if token == 'and' or token == 'or' or token == 'not':
+                pending_operation = token.upper()
+            else:
+                results = set()
+                pointers = self.index.get(token)
+                if not pointers:
+                    continue
+                i = 0
+                for pointer in pointers:
+                    comment = CsvHelper.read_comment(pointer)
+                    results.add(comment)
+                    if i > number_of_results:
+                        break
+                    i += 1
+                result_stack.append(results)
+
+                if pending_operation != '':
+                    results = getattr(self, pending_operation)(result_stack.pop(), result_stack.pop())
+                    result_stack.append(results)
+
+        return result_stack[-1]
+
+    def AND(self, left_bag, right_bag):
+        return left_bag.intersection(right_bag)
+
+    def NOT(self, left_bag, right_bag):
+        return left_bag - right_bag
+
+    def OR(self, left_bag, right_bag):
+        return left_bag.union(right_bag)
+
+    def NOT(self, left_bag, right_bag):
+        return right_bag - left_bag
```

Comment:
this makes no sense. 1.: It is a duplicate of line 60, which could be a problem because you are calling the methods by name. 2.: NOT is used before the second query token, which means the second results should be subtracted from the first ones, like it is in line 60.

Reply:
ups..